langchain_chat.py

```diff
@@ -0,0 +1,75 @@
+import streamlit as st
+from streamlit_chat import message
+from langchain.embeddings.openai import OpenAIEmbeddings
+from langchain.chat_models import ChatOpenAI
+from langchain.chains import ConversationalRetrievalChain
+from langchain.document_loaders.csv_loader import CSVLoader
+from langchain.vectorstores import FAISS
+import tempfile
+
+user_api_key = st.sidebar.text_input(
+    label="#### Your OpenAI API key 👇",
+    placeholder="Paste your openAI API key, sk-",
+    type="password")
+
+uploaded_file = st.sidebar.file_uploader("upload", type="csv")
+
+if uploaded_file :
+   #use tempfile because CSVLoader only accepts a file_path
+    with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+        tmp_file.write(uploaded_file.getvalue())
+        tmp_file_path = tmp_file.name
+
+    loader = CSVLoader(file_path=tmp_file_path, encoding="utf-8", csv_args={
+                'delimiter': ','})
+    data = loader.load()
+
+    st.write(data)
+
+
+embeddings = OpenAIEmbeddings()
+vectorstore = FAISS.from_documents(data, embeddings)
+
+chain = ConversationalRetrievalChain.from_llm(
+llm = ChatOpenAI(temperature=0.0,model_name='gpt-3.5-turbo'),
+retriever=vectorstore.as_retriever())
+
+def conversational_chat(query):
+    
+    result = chain({"question": query, 
+    "chat_history": st.session_state['history']})
+    st.session_state['history'].append((query, result["answer"]))
+    
+    return result["answer"]
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The function `conversational_chat` modifies the session state directly. It would be better to separate the concerns of updating the session state and generating the chat response. This would make the code easier to test and maintain.
```diff
- def conversational_chat(query):
-     
-     result = chain({"question": query, 
-     "chat_history": st.session_state['history']})
-     st.session_state['history'].append((query, result["answer"]))
-     
-     return result["answer"]
+ def conversational_chat(query, history):
+     result = chain({"question": query, "chat_history": history})
+     return (query, result["answer"])
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->